### ISSUE TITLE
Have one manufacturing queue for entire settlement 

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#1236]
**Is your feature request related to a problem? Please describe.**
- Currently, there is no queue for manufacturing.
- When a lane or slot is available, the settlement will compute what process to undertake
- If there are more than one lander hab, workshop, manufacturing shed, machinary hab, players have to open all of them to discover what the settlement is manufacturing

**Describe the rationale**
- Player commander should be given better decision-making authority to decide what to manufacture. 

**Describe the goal(s) you have in mind**
- Create a Manufacturing tab in Monitor Tool to show what the settlement will manufacture
- Show the rating score on each process
- Allow manually adding and deleting manufacturing processes
- Add a manufacturing queue for storing future manufacturing processes
- Allow players to click to increase or decrease priority on a manufacturing process

**Specifications**
- post 3.7.0 build 9292

**Additional context**
- May implement these features in line with https://github.com/mars-sim/mars-sim/issues/1062

**Comment by mokun** [comment]
@bevans2000 

In a lot of city building games, there's one central construction queue that shows what is being built.

We could in principle create one single queue panel that let player see all the manufacturing processes that are running among Workshops, Lander Hab, Machinery Hab, etc.

This allows player to decide if no overlapping effort in making the same part or resources among various buildings is warranted.

**Comment by bevans2000** [comment]
Sure. I like this idea.

**Comment by bevans2000** [comment]
What we need is a new class called ProductionManager that can maintain the queue and act as an overseeing for the various production functions. Then use a ScheduledEvent to periodically add to the production queue. Should be a nice new feature.
I'm struggling for time at the moment because of work and working on #1285 which is good because I am doing lots of new UnitTest for the Tasks.
Also drop the ability to creation new processes for individual Buildings in the UI but have it centralised. The metatask logic will be much easier as it just acts as a proxy to the ProductionManager for the Settlement being checked.

**Comment by mokun** [comment]
For now, I'm going to simplify add the ability to queue in each building that has manufacturing function. 

So far, if a workshop supports only 2 concurrent processes and there are already 2 on-going processes. Player cannot add the 3rd one. 

This is very inconvenient since player has to wait for those 2 processes to get done first before he can add one. 

Also, if the settlement decides to add 2 processes right after the previous 2 was done. The players has to wait again.

Adding a queue now for each building makes sense.

In future, we'll consider a settlement-wide queue.

**Comment by bevans2000** [comment]
Would a Settlement queue make more sense and be more realistic?

**Comment by mokun** [comment]
> Would a Settlement queue make more sense and be more realistic?

The short answer is yes.

A player commander would be goal-oriented. 

He wouldn't care which building can manufacture an item as long as it gets done.

**Comment by mokun** [comment]
@bevans2000 

With regard to creating a centralized manufacturing queue, here's the issue thread you may be looking for.

**Comment by bevans2000** [comment]
What's needed is a new ```ProcessManager``` that is responsible for queueing and creating new processes. The creation logic currently in the Manufacture class would be moved. When a Manufacture completes a process it would inform the  manager and that would assign the best choice off the queue. Essentially it acts as a controller for all the Manufacture instances in a settlement.
The manager would takes over automatically adding items to the queue.

**Comment by mokun** [comment]
> ProcessManager that is responsible for queueing and creating new processes

So will this ProcessManager works for FoodProduction as well ?

**Comment by bevans2000** [comment]
> > ProcessManager that is responsible for queueing and creating new processes
> 
> So will this ProcessManager works for FoodProduction as well ?

Food production is a bit different I think. Salvage & Manufacturing have a lot of synergy.

**Comment by mokun** [comment]
One question is to determine optimally how many person to work on a process in a building.

It is not a bad idea to share the work load.

It shouldn't be a solo task like right now.

For instance, it's natural for a number of mechanics to work on the same vehicle in a garage to speed up the job.

Or for a number of cooks to work in a kitchen to prepare for a meal.

**Comment by bevans2000** [comment]
Yes, get the current approach working. That is actually focussing on assigning new processes.
Your question is actually about what happens after a process has been assigned. So I think it is a separate issue that follows on. However, I don't see it as a big problem. If each active process creates a Settlement task, then that support multiple Workers accepting

**Comment by bevans2000** [comment]
Right, I've had a look at this. We cannot do this without also addressing the Salvage processes as they share the same workshop. My thoughts are:

- Create a ```ManufacturingManager``` that control all jobs for the Manufacture workshops in a single Settlement.
- Maintains a single queue of processes waiting to be run; both Manufacture & Salvage
- A new ```ManufactureMetaTask``` that is a SettlementTask and replaces the 2 existing MetaTasks.
- The new MetaTask creates SettlementTasks covering effort for running Processes and any capacity to start new ones

It's quite a wide ranging change but will be a lot more efficient. The current code base looks expensive in terms of processing.

**Comment by mokun** [comment]
> Secondly ManufactureConstructionMaterial is the same logic using a subset of processes. This can be removed and combined into the single queue.

Guess we can even roll in food production processes into this new logic.

> It's quite a wide ranging change but will be a lot more efficient. The current code base looks expensive in terms of processing.

It used to be even more expensive because the meta tasks were querying for the quantity. It has since been tone down a while back.

**Comment by bevans2000** [comment]
Food needs to be separate because it uses a different Function. Manufacturing, Salvage and ConstructionMaterial Task all use the Manufacture Function so they can go together

**Comment by mokun** [comment]
> Food needs to be separate because it uses a different Function. Manufacturing, Salvage and ConstructionMaterial Task all use the Manufacture Function so they can go together

Note that Food production is a very close clone of Manufacturing when I first created this concepts a number of years ago.

On the other hand, I agree that meal and dessert preparation is an entirely different animal.

**Comment by bevans2000** [comment]
Sure, this could still be expensive in terms of processing. My approach is only check inputs for processes on the queue. 
I'm going to disable salvage all together and get manufacturing working with the user adding queue items.
Then work on how to automatically add items to the queue as demand requires

**Comment by mokun** [comment]
> I'm going to disable salvage all together and get manufacturing working with the user adding queue items.

Ok. We don't really want to see a settlement salvaging any robots and vehicles.

It's possible that some containers may be in excess, if amount resources are dwindling. But it's very hard to determine if it should salvage all the bags, because in doing so, it will cause insufficient containers and only come back and hurt the dig local tasks and collect resource missions and prevent them from getting started.

**Comment by bevans2000** [comment]
So bit of an update on the approach.

- Create a new ```ManufacturingManager``` that maintains a quete of Manufacturing and Salvage processes. This is specific to a single Settlement.
- Change the Settlement Unit Window so it shows the queue as well as what is processing
- The manager will decide what new processes to add automatically to the queue, check once per sol
- Manufacturing function has no queue anymore or ability to choose processes, it's a passive entity now
- A new SettlementMetaTask class called ```ManufacturingMetaTask``` that create a single backlog task for any Manufacturing building that has capacity or work in progress. This will cover adding new processes. 
- Manufacturing manager will control the allocation of new processes based on the priority queue and resources available
- Remove the existing Salvage and Manufacturing metattask classes
- Also I see no need for the Manufacture Construction materials task, it replicates what we already have

This will need a gradual change and I'll do it in multiple commit son a branch obviously. The hardest is deciding how to auto select process to add to the queue. Needs to be based on demand. What do you think is the answer ?

**Comment by mokun** [comment]
> Also I see no need for the Manufacture Construction materials task, it replicates what we already have

I see this is an important design choice here.

How do we consider how cement, concrete, bricks, etc are created ?

Do you imagine a super manual process, or a totally automated process (like resource processes) ?

Construction materials will become very useful as soon as we define how we want construction missions to proceed under certain overarching objectives.

How much bottleneck can we accept when making Construction materials ?

**Comment by mokun** [comment]
> The hardest is deciding how to auto select process to add to the queue. Needs to be based on demand.

So we want to base on the value point instead, or just the demand ? 

Value point takes into consideration the supply. 

Also, say process A are available in both workshop 1 and workshop 2, we should allow certain preferences to be integrated into the priority score calculations now, rather than later.

For now, since settlements aren't that big. Location choice is not relevant. But when there are more than 1 zone in future, each zone should have its own manufacturing objectives that a player commander can customize and change on the fly.

Can we add user preference to murliply  the manufacturing priority score ? It will computed initially based upon the settlement and sponsor objectives.

We already have a tab for editing and showing the preferences for mission, tasks, science, etc.

Now it's time we add manufacturing preferences.


But add the

**Comment by mokun** [comment]
On top of the preferences, I see we can add the concept of commander-based focus. 

If I'm a commander, I should be able to manually designate or single out a workshop to specifically manufacture one or two items or resources for me. 

This workshop will be dedicated to manufacturing just that list of processes . 

Then if it lacks the input resources, it should flag that process as problematic and offers some feedbacks in commander's report and log the issue. 

This way, player would be notified and he may further troubleshoot the chain of events or root cause. 

We should have a option checkbox to allow that workshop to go idle or to be included back into the settlement-based fold.

**Comment by mokun** [comment]
If a workshop is under the commander-based mode, then it will go back to the old metatask method of allowing a settler to rate how capable he is to perform that task and choose it among many other metatasks.

**Comment by bevans2000** [comment]
Hmmm, OK the automatic based logic on value or demand makes sense. That will fit with what I'm doing. I have just done a commit on a branch with the basics implemented. Check it our and have a play with it. I'm going to work on a couple of tidy up/simplifications next before tackling Salvaging. Then at the end revisit the automatic creation; the latter I'll do in a separate branch. So plan to have a full manual solution on the master branch soon.
Overall I think this is going to be a great set of changes and make it a lot clearer and easier to understand. Also should be more efficient.

As to the Commander mode. I think we need to think of it in a different way. The Commander will have the ability to explicitly remove individual items from the central Queue and assign them directly to a Workshop. That can be an expansion of the new solution.

**Comment by mokun** [comment]
What about implementing manufacturing process preference score ?

**Comment by bevans2000** [comment]
You mean preferences down to the individual process names? But automatic selection should be based on the Settlements needs?

**Comment by mokun** [comment]
> automatic selection should be based on the Settlements needs?

Yes it's still based on the current demand for each settlement. But each settlement/sponsor has pre-conceived notion of what it needs that should be influenced by the the overarching objective of that settlement.

**Comment by bevans2000** [comment]
@mokun I've done the changes to allow SalvageProcessInfo to be added to the new single queue with a much better UI.
But I've thought of a radical idea to significantly simplify things. I have noticed that the vast majority of Salvage processing and Manufacturing logic is the same; a lot of duplicate code. So I proposing to create a new Task that replaces the ```ManufactureGood``` and ```SalvageGood``` called ```WorkshopTask```. This is a new Task that moves a process through the working time logic through to completion. This new task would work on a ```WorkshopProcess``` that contains all the common attributes of the ```SalvageProcess``` and ```ManufactureProcess```. 
The benefits is there is no difference between the running on the 2 different process types and hence can simplify the ```Manufacture``` class as it just a common set of processes active. It also virtually halves the amount of code involved.
Secondly; currently having 2 difference process type completing for the same Entity, Manufacture, introduces a lot of complexity.
What do you think?

**Comment by mokun** [comment]
I know each manufacturing process requires one concurrent 3D printer to be online to proceed. 

Should a salvage process require a space to proceed? It certainly doesn't need a 3D-printer to take apart something.

That's the difference between a manu process and a salvage process.

**Comment by bevans2000** [comment]
I could complete the single queue approach using modified versions of the existing Tasks. Then do the above as a new Issue.
I have manually added processes working. Then look at doing the automatically adding to the queue.

**Comment by bevans2000** [comment]
Ok, that's a small difference we can handle, the root of this is the difference between processing time for the 2 types. Salvage doesn't have processing but Manufacturing does. What's interesting is we could develop different tools, not just a 3d printer. We could have a lathe for metalwork.
But there is a limit on the number of Processes a Manufacturer instance can run concurrently, that applies to both types. I think this is doable and would be a good change. Going to have more of a think on it.

**Comment by bevans2000** [comment]
@mokun I am going to look at automatically adding processes to the queue this weekend. This will be based on the value of the Outputs, similar to the current approach but much more cutdown. I am thinking that the user can control

- how many get added per sol
- the minimum value of the output

Processes are added to the queue based on these creation, same logic applies to Salvage as well. The controls will be Settlement Preferences and use the existing logic for that.

**Comment by bevans2000** [comment]
@mokun Here is what the new Manufacturing tab looks like. You can see that the user first selects the Output required and secondly selects the Process that generates it. This show the new parameters for the automatic creation.
![Screenshot 2025-01-11 220516](https://github.com/user-attachments/assets/f125b1f1-c962-41ce-b555-488d6c7e90f7)

**Comment by mokun** [comment]
That's nice. 

Q: how many processes would you allow to be under the 'Queue' tab ?  

I feel like it's shouldn't be infinite. 

What determine how large the queue is ?

**Comment by bevans2000** [comment]
I've set it 10 by default but it is controlled by a Preference setting so under user control.

I plan to change the value score to use our Rating Score class so we can break it down.

I will try and create a PR today so you can take into the master to try it out. Or just jump on the branch

**Comment by mokun** [comment]
> I've set it 10 by default 

ic

> but it is controlled by a Preference setting so under user control.

great !

Q: will the queue be autogenerated ? 

I can see that autogenerating a queue of processes has the advantage of showing the player VISUALLY what kind of priority the settlement has in terms of manufacturing.

If I see the queue has 10 "Make concrete" processes, then I know the settlement is very interested in constructing new buildings.

And in that case, if i don't like the look of it, I will intervene and tune up some parameters to the degree that it would only autogenerate 5 "Make concrete" processes, leaving room for 5 other processes.

**Comment by bevans2000** [comment]
@mokun Give the new approach a try please. Currently; it only adds one instance of a Process; so it will not create 10 Concrete processes. I think that is correct.
The selection is heavily controlled by what a Settlement can do in terms of Person Skill and Tech Level; both of these control what is offered to the user and automatically added.
What I do like in this implementation is the Value of the output to the Settlement is used as a priority; this is exactly the correct approach. A user can influence the priority by applying a bonus modifier that will  inflate or deflate the Value. The logic is there but I've got to do the UI yet; that is next.
I don't want to make a new tab in the Monitor tool as I feel this is to specific; the newly redesigned Manufacture tab panel works well.
I love the new approach of the user selecting the desire output first and then an appropriate Process; makes it mush easier to use
Auto adding Salvage processes needs some thoughts; need a way to identify if a Settlement has too many of something. If it meets that criteria; then the existing logic can be used. Need to work out how to assess "too many barrel".
